### PR TITLE
Make it possible to void invoices

### PR DESCRIPTION
**Pull request by rluodev**
This adds the ability to void an invoice and adds sorting options on both admin- and user-side to sort and view these voided invoices.

Once merged, this will close #4711.

Voided invoice

Screenshot 2023-12-13 at 9 01 18 AM
paid invoice with void button disabled

Screenshot 2023-12-12 at 5 25 35 PM
sent invoice with void button

Screenshot 2023-12-12 at 5 25 54 PM

Screenshot 2023-12-12 at 5 23 52 PM

Screenshot 2023-12-12 at 5 26 56 PM

### CODE REVIEW

**Comment by sarthaktexas** [comment]
by rawr???

**Comment by rluodev** [comment]
> by rawr???

that's just what I could think of

**Review by sarthaktexas — CHANGES_REQUESTED** [review]
Could I get captions on the images in the PR description? Specifically, the first few because I cannot tell the difference.

**Comment by sarthaktexas** [comment]
Also, I cannot approve this until I test it locally, so you'll have to wait until tomorrow for an approval review. I've reviewed your code and have some comments. I will submit a review tomorrow when I have a pair of fresh morning eyes. I love the feature!

File: app/views/hcb_codes/_invoice.html.erb
Comment (sampoder) [inline app/views/hcb_codes/_invoice.html.erb]:
```suggestion
    <%= link_to "Void this invoice",
      invoice_void_path(@invoice.id),
      disabled: !organizer_signed_in? || @invoice.voided? || @invoice.paid_v2?,
      method: :post,
      data: { confirm: "Are you sure you want to void this invoice? This action cannot be undone." },
      class: "btn bg-error mr1 right" unless @invoice.paid_v2? || @invoice.voided?  %>
```

Let's hide this button if the invoice is paid or already voided, reduces clutter imo.

File: app/views/hcb_codes/_invoice.html.erb
Comment (sampoder) [inline app/views/hcb_codes/_invoice.html.erb]:
```suggestion
      data: { confirm: "Are you sure you want to void this invoice? Voided invoices can not be paid and this action cannot be undone." },
```

Let's be super clear with folks about what this button does.

File: app/views/hcb_codes/_invoice.html.erb
Comment (sampoder) [inline app/views/hcb_codes/_invoice.html.erb]: Can an invoice be both voided and archived? It seems like I can still archive voided invoices. Do we want to support this?

File: app/models/invoice.rb
Comment (sampoder) [inline app/models/invoice.rb]: Let's continue using `void_v2?`. Afaik `mark_void!` should lead to `void_v2?` returning true.

File: app/controllers/admin_controller.rb
Comment (sampoder) [inline app/controllers/admin_controller.rb]:
```suggestion
    relation = relation.open_v2 if @open
```
When we void a invoice they should transition to the `void_v2` state and therefore not be open. Were you observing that when testing this?

File: app/controllers/admin_controller.rb
Comment (sampoder) [inline app/controllers/admin_controller.rb]:
```suggestion
    relation = relation.void_v2 if @voided
```

File: app/controllers/invoices_controller.rb
Comment (sampoder) [inline app/controllers/invoices_controller.rb]:
```suggestion
      relation = relation.void_v2
```

**Review by sampoder — COMMENTED** [review]
Couple of comments from me. Main thing is that I think we should be using `.void_v2` instead of creating a separate `.voided`.

File: app/controllers/admin_controller.rb
Comment (rluodev) [reply app/controllers/admin_controller.rb]: Yeah, while testing, I had to use .excluding in order to get it to actually exclude voided ones. not sure why.

File: app/views/hcb_codes/_invoice.html.erb
Comment (rluodev) [reply app/views/hcb_codes/_invoice.html.erb]: I mean it doesn't change anything visually aside from the invoice archive button changing between archive and unarchive, so I think we can just hide the archive button once voided? Will push a change for this.

File: app/models/invoice.rb
Comment (rluodev) [reply app/models/invoice.rb]: Done!

**Comment by rluodev** [comment]
@sampoder re: the void_v2 thing - it only works for *some* invoices, apparently, and the voided? call is a much more reliable way to keep track of voided status since we keep a record on our end of when the invoice was voided ~~instead of asking stripe, "hey is this void" and maybe not receiving back a correct response.~~

wait no i totally assumed that void_v2 was a call to stripe but anyway it still doesn't always work.

File: app/controllers/admin_controller.rb
Comment (sampoder) [reply app/controllers/admin_controller.rb]: Were you running `.mark_void!`?

**Comment by sampoder** [comment]
@Soar1826 how are you testing this locally? I'd like to pull it up on my computer and investigate.

File: app/controllers/admin_controller.rb
Comment (rluodev) [reply app/controllers/admin_controller.rb]: yep, 
Screenshot 2023-12-13 at 11 08 10 AM

**Comment by rluodev** [comment]
wait no nvm it was a weird thing that happened to one invoice on my end

**Comment by rluodev** [comment]
void_v2 works fine

**Comment by sampoder** [comment]
Oh hell yeah!

File: app/controllers/admin_controller.rb
Comment (sampoder) [inline app/controllers/admin_controller.rb]:
```suggestion
    relation = relation.open_v2 if @open
```
I believe it's implied that `open_v2` invoices are not `void_v2`

File: app/services/invoice_service/void.rb
Comment (sampoder) [inline app/services/invoice_service/void.rb:1]: In the same way we have `MarkPaid`, let's have `MarkVoid` instead of just `Void`.

File: app/controllers/invoices_controller.rb
Comment (sampoder) [inline app/controllers/invoices_controller.rb]: Would it make sense to move this into the service? Feels more in line with what we do on `MarkPaid`

File: app/views/hcb_codes/_invoice.html.erb
Comment (sampoder) [inline app/views/hcb_codes/_invoice.html.erb]: I could archive an invoice and then void it, should we display that in any way?

File: app/views/hcb_codes/_invoice.html.erb
Comment (sampoder) [inline app/views/hcb_codes/_invoice.html.erb]:
```suggestion
    <%= link_to "Void this invoice",
      invoice_void_path(@invoice.id),
      disabled: !organizer_signed_in? || @invoice.void_v2? || @invoice.paid_v2?,
      method: :post,
      data: { confirm: "Are you sure you want to void this invoice? Voided invoices can not be paid and this action cannot be undone." },
      class: "btn bg-error mr1 right"unless @invoice.void_v2? || @invoice.paid? %>
  <%= link_to (@invoice.archived? ? "Un-archive" : "Archive"),
    (@invoice.archived? ? invoice_unarchive_path(@invoice) : invoice_archive_path(@invoice)),
    disabled: !organizer_signed_in?,
    method: :post,
    class: "btn bg-muted" unless @invoice.void_v2? %>
```

Test this rq but I think this is valid and imo cleaner

**Review by sampoder — CHANGES_REQUESTED** [review]
Couple from me! Looking really good.

File: app/views/hcb_codes/_invoice.html.erb
Comment (rluodev) [reply app/views/hcb_codes/_invoice.html.erb]: The general idea is that voiding takes precedence - you can't archive after voiding, so it makes no sense to show something as archived when the void tab does the same thing and only shows voided like how archived only shows archived.

File: app/views/hcb_codes/_invoice.html.erb
Comment (rluodev) [reply app/views/hcb_codes/_invoice.html.erb]: I feel like we do want to hide the button to archive as well when voided, doesn't really make sense to void and then archive.

File: app/views/hcb_codes/_invoice.html.erb
Comment (sampoder) [reply app/views/hcb_codes/_invoice.html.erb]: That makes sense to me; thanks for clarifying! Just to sanity check, after we "void" an archived invoice does it show up as "archived" anywhere?

File: app/services/invoice_service/void.rb
Comment (rluodev) [reply app/services/invoice_service/void.rb:1]: done!

File: app/views/hcb_codes/_invoice.html.erb
Comment (sampoder) [reply app/views/hcb_codes/_invoice.html.erb]: Oh yes, that's my bad! I intended for that to be the case. Let me quickly fix my suggestion.

Comment (sampoder) [reply app/views/hcb_codes/_invoice.html.erb]: Fixed, sorry

File: app/controllers/invoices_controller.rb
Comment (rluodev) [reply app/controllers/invoices_controller.rb]: I was doing it in-line with the other actions in invoice controller, like archive, etc.

File: app/views/hcb_codes/_invoice.html.erb
Comment (rluodev) [reply app/views/hcb_codes/_invoice.html.erb]: thanks! also, one more thing: shouldn't `unless @invoice.void_v2? && @invoice.paid?` be `unless @invoice.void_v2? or @invoice.paid?`?

File: app/views/hcb_codes/_invoice.html.erb
Comment (rluodev) [reply app/views/hcb_codes/_invoice.html.erb]: No, the status are all listed as void I believe, will test tho.

Comment (rluodev) [reply app/views/hcb_codes/_invoice.html.erb]: yep, they all show as void.

Comment (sampoder) [reply app/views/hcb_codes/_invoice.html.erb]: Perfect!

File: app/views/hcb_codes/_invoice.html.erb
Comment (rluodev) [reply app/views/hcb_codes/_invoice.html.erb]: @sampoder

Comment (sampoder) [reply app/views/hcb_codes/_invoice.html.erb]: Yes, though use `||`

File: app/controllers/invoices_controller.rb
Comment (sampoder) [reply app/controllers/invoices_controller.rb]: Let's move it into the service, that way we can reuse the service and it'll be consistent

**Review by sampoder — APPROVED** [review]
Awesome, LGTM!

File: app/models/invoice.rb
Comment (garyhtou) [inline app/models/invoice.rb]: Technically using `voided_at` works! However, when using AASM, this code pattern of using `*_at` to determine state can be buggy. Here's why!

Let's say an Invoice goes from `open` -> `paid`. Each state transition usually involves setting a timestamp (e.g. `opened_at`, and `paid_at`). We'd expect `where.not(opened_at: nil)` to give us a list of `open` invoices, however, this actually gives us a list of invoices that were _open at some point_. This is because state transitions do not clear the timestamps of other states.

In this specific case, it works fine because `void` is a terminal state. Once it's voided, there is no other state it could transition to.

To prevent possible bugs like these in the future, let's switch it to using the `aasm_state` column for determining state.

```suggestion
  scope :voided, -> { where(aasm_state: "void_v2") }
```

---

Also, I just wanted to mention that AASM automatically provides scopes. In this case, our `voided` scope is the exact same as AASM's provided `void_v2` scope.

File: app/views/hcb_codes/_invoice.html.erb
Comment (garyhtou) [inline app/views/hcb_codes/_invoice.html.erb]: Since we have `unless @invoice.void_v2? || @invoice.paid?` on line 153, I think it might not be needed

```suggestion
      disabled: !organizer_signed_in?,
```

File: app/views/hcb_codes/_invoice.html.erb
Comment (garyhtou) [inline app/views/hcb_codes/_invoice.html.erb]: I see that we're also checking for `deposited?` on line 51. I'm actually not sure if this is necessary; but imma assume it is 🤷‍♂️ (it's tough to say because the state management in this model needs some work)

```suggestion
      class: "btn bg-error mr1 right" unless @invoice.void_v2? || @invoice.deposited? || @invoice.paid? %>
```

File: app/views/hcb_codes/_invoice.html.erb
Comment (garyhtou) [inline app/views/hcb_codes/_invoice.html.erb]: It seems like the intention is to not allow `void` -> `archive`. However, at the moment, it's possible to `void` a previously archived invoice (and it'd be in both states!). Is that intentional?

The void button is visible for archived invoices.

File: app/services/invoice_service/mark_void.rb
Comment (garyhtou) [inline app/services/invoice_service/mark_void.rb]: Three edits:
1. Wrap in a database transaction. If anything fails, it will rollback changes; this way invoices aren't partially voided.
2. use `close_stripe_invoice` which is already defined. You'll need to make `close_stripe_invoice` a public method. If you'd like, feel free to give it a better name (e.g.`void_stripe_invoice`)
3. Use `save!`. If there is a validation error, I'll rollback the changes and show an error to the User

```suggestion
      Invoice.transaction do
        invoice.voided_at = DateTime.now
        invoice.voided_by = @user

        invoice.close_stripe_invoice
        invoice.mark_void!

        invoice.save!
      end
```

File: app/controllers/invoices_controller.rb
Comment (garyhtou) [inline app/controllers/invoices_controller.rb]: We should update the `unpaid` and `archived` scope rather than excluding voided invoices here

**Review by garyhtou — CHANGES_REQUESTED** [review]

File: app/views/hcb_codes/_invoice.html.erb
Comment (rluodev) [reply app/views/hcb_codes/_invoice.html.erb]: Yep, this is intentional, as the invoice can be possibly archived and then voided, at which point it will just show up solely in the voided tab

File: app/controllers/invoices_controller.rb
Comment (rluodev) [reply app/controllers/invoices_controller.rb]: fixed!

**Review by sarthaktexas — APPROVED** [review]
lgtm great job

File: app/services/invoice_service/mark_void.rb
Comment (garyhtou) [inline app/services/invoice_service/mark_void.rb]: Let's do `mark_void!` before `close_stripe_invoice`. This is because we can rollback `mark_void!`. But we can't rollback `close_stripe_invoice` since it's an API call to Stripe

```suggestion
        invoice.mark_void!
        invoice.close_stripe_invoice
```

File: app/services/invoice_service/mark_void.rb
Comment (garyhtou) [inline app/services/invoice_service/mark_void.rb]: Add `timestamp: true` to the `aasm do` block in the model. AASM will then automatically set the timestamp.

https://github.com/aasm/aasm#timestamps




You can also add a parameter to the AASM transition to take in a user.
See this as an example: https://github.com/hackclub/bank/blob/945eadcce352b789331b3058a8fe1de55e8c727a/app/models/disbursement.rb#L117-L123

**Review by sarthaktexas — APPROVED** [review]

File: app/controllers/invoices_controller.rb
Comment (garyhtou) [inline app/controllers/invoices_controller.rb]:
```suggestion
    ::InvoiceService::MarkVoid.new(invoice_id: @invoice.id, user: current_user).run
```

File: app/views/hcb_codes/_invoice.html.erb
Comment (garyhtou) [inline app/views/hcb_codes/_invoice.html.erb]: i _think_ this does the same? i'm not confident about this change
```suggestion
    <% if @invoice.open_v2? %>
```

File: app/views/hcb_codes/_invoice.html.erb
Comment (garyhtou) [inline app/views/hcb_codes/_invoice.html.erb]: maybe this works
```suggestion
      class: "btn bg-error mr1 right" if @invoice.open_v2? %>
```

**Review by garyhtou — COMMENTED** [review]
looks good, just a couple small things then good to merge

**Pull request merged**
